Refactor WindowHandle and image transformation.

de-vri-es opened the pull request:
This PR makes access through a window handle `O(1)` instead of doing a lookup each time. Since access is now cheap, the PR also directly exposes a user transform which will be applied to the image and overlays. Then there's some misc API tweaks.

de-vri-es commented:
@vsuryamurthy: Do you want to take a look?

vsuryamurthy commented on src/backend/context.rs:
```diff
@@ -821,23 +673,24 @@ impl Context {
 			} else {
 				let context_handle = ContextHandle::new(self, event_loop);
-				let mut window_handle = WindowHandle::new(context_handle, event.window_id());
+
```
Mistake I assume.

de-vri-es replied:
The extra newline? Yeah, that shouldn't be there.

de-vri-es replied:
Removed it :)

vsuryamurthy commented on src/backend/window.rs:
```diff
@@ -58,19 +63,44 @@ pub struct WindowHandle<'a> {
+		&self.context().windows[self.index]
+	}
+
+	/// Get a mutable reference to the window.
+	fn window_mut(&mut self) -> &mut Window {
+		let index = self.index;
+		unsafe { &mut self.context_mut().windows[index] }
```
Although I see the need for mutable reference to the window, couldnt you potentially also destroy the windows as well?

de-vri-es replied:
No, you can't drop an object through a mutable reference :)

vsuryamurthy commented on src/backend/context.rs:
```diff
@@ -406,40 +282,44 @@ impl Context {
 		event_loop: &EventLoopWindowTarget,
```
Not specific to this line. 

I am curious why the structure is the following:
impl Context {}
impl ContextHandle {}
impl Context{}

Why not have 
impl Context { `every function here`}
impl ContextHandle{}

de-vri-es replied:
Hmm, this is a relic from back when I still planned to expose `Context` directly to users of the crate. I put all the public APIs at the top and then the internal stuff at the buttom. But now `Context` itself is private, so it doesn't make much sense anymore.

I'll move it around after this PR :)

vsuryamurthy reviewed (commented):
The first commit kind of confused me. The last line goes like "That also means you can not get a `&mut ContextHandle` from a
`&mut WindowHandle` anymore." So I was thinking why not just remove the mutability of WindowHandle reference and it seems like you are doing that. Am I right?

de-vri-es commented on src/backend/context.rs:
```diff
@@ -792,7 +644,7 @@ impl Context {
 		// That's not allowed, of course, so temporarily swap the event handlers with a new vector.
 		// When we've run all handlers, we add the new handlers to the original vector and place it back.
 		// https://newfastuff.com/wp-content/uploads/2019/05/dVIkgAf.png
```
I think this is still accurate. Why would it need to be changed?

vsuryamurthy replied:
Again my bad. replace literally depicted the comment as you have two arguments, event handler and a new vector. While take is a bit more esoteric but I guess does the same job. Hence my confusion.

de-vri-es replied:
Ah, that was a suggestion from clippy. But indeed, it does exactly the same thing :)

de-vri-es commented on src/backend/util/uniforms_buffer.rs:
```diff
@@ -1,5 +1,13 @@
 use super::buffer::create_buffer_with_value;
 
+pub unsafe trait ToStd140 {
```
Because the trait implies that `Output` is layout compatible with the relevant struct from the shader. If it is not, it will cause the shader to access the wrong memory.

Although I'm not 100% sure if that can cause memory safety issues for the Rust program, if I'm honest.

vsuryamurthy replied:
Okay interesting!

de-vri-es commented on src/backend/window.rs:
```diff
@@ -79,63 +106,197 @@ impl<'a> WindowHandle<'a> {
 	///
-	/// Any subsequent operation on the window will return [`InvalidWindowId`].
-	pub fn destroy(&mut self) -> Result<(), InvalidWindowId> {
-		self.context_handle.destroy_window(self.window_id)
```
Probably not. But didn't I remove it?

vsuryamurthy replied:
Sorry my bad. You have removed it already. I thought I saw a destroy_window but it is part of Context not ContextHandle.

de-vri-es commented:
> The first commit kind of confused me. The last line goes like "That also means you can not get a `&mut ContextHandle` from a
> `&mut WindowHandle` anymore." So I was thinking why not just remove the mutability of WindowHandle reference and it seems like you are doing that. Am I right?

Not exactly. Previously, `WindowHandle` had a mutable reference to a `ContextHandle` and a window ID. Then each function actually looked up the window by ID again, and all of them could fail with `InvalidWindowId`.

The window ID is now replaced with an index in the window vector. This means access to the window is O(1) instead of the O(n) it was before, and that functions can no longer fail to look up the window. But it also means we're not allowed to modify the window vector, because that would invalidate the index.

Internally we still need a mutable reference to the `ContextHandle`, because we need to get a mutable reference to the window to modify event handlers and properties. Internally we also still *can* get a mutable reference to the `Context`. However, users of the crate can't get a mutable reference to the `ContextHandle` anymore, because that would let them modify the window vector and invalidate the index.

The only way to get a `&mut ContextHandle` is by *consuming` the `WindowHandle`, but you can't do that with a `&mut WindowHandle`. That's also why I updated the event handlers to get a `WindowHandle` by value, so they can consume the handle.

vsuryamurthy commented:
Once the WindowHandle is consumed, it is basically destroyed. So every update to event handlers or any function that consumes WindowHandles, requires a WindowHandle then right? I mean do you keep creating WindowHandle s for every time we need to do something?

de-vri-es commented:
> Once the WindowHandle is consumed, it is basically destroyed. So every update to event handlers or any function that consumes WindowHandles, requires a new WindowHandle then right? I mean do you keep creating WindowHandle s for every time we need to do something?

Yes :)

This is still a lot more efficient then before, because then every *use* of a window handle would do the lookup. Now it's only when creating the handle.

vsuryamurthy reviewed (approved):
LGTM
